swarm/shed/index.go

```diff
@@ -78,6 +84,11 @@ type Index struct {
 	decodeKeyFunc   func(key []byte) (e IndexItem, err error)
 	encodeValueFunc func(fields IndexItem) (value []byte, err error)
 	decodeValueFunc func(value []byte) (e IndexItem, err error)
+
+	// triggers are used for signaling
+	// subscriptions to continue iterations.
+	triggers   map[uint64]chan struct{}
```

Comment:
a trigger should be associated with a subscription not an index.
E.g., stream pkg subscribe to bin

Reply:
In this code, sending and receiving part of the subscriptions are completely decoupled. Subscription only receives data and controls whether the receiving should stop or terminate. While only index is responsible for writing (putting keys and values) and signaling any subscriptions on such events. So the responsibility is completely divided. Index does not need to track every existing subscription to trigger them, but just needs to call one function, while any subscription can be created and stopped without the need to adjust any trigger invocation.

It is of course possible to have triggers on Subscription, but then, both sending and receiving parts need to be changed when a subscription is created or stopped. Which may not be the problem if we do not create subscriptions dynamically or we want to manage them outside of this package.

For subscribing to a particular bin which is a prefix in front of the timestamp in the index, we would need subscriptions that work on a prefix and to see if we want to manage this subscriptions in general way in shed package or very specific in localstore.

Reply:
@janos as long as it is possible to have bin specific triggers i am ok with this.

Reply:
but looking at the code i dont this that is possible. Even if trigger is created in `NewSubscription` it should be assigned as a field of the subscription so that it can be triggered by the implementation. But it seems to me to be better to just give the channel as an argument. That way it could then just be a lazy event subscription triggered (written to with a default select) if there is a new item relevant for the iteration. Otherwise how would you do this?